### PR TITLE
Split travis tests in bigger matrix

### PR DESCRIPTION
**Pull request by MaxThevenet**
As we were getting closer to the timeout on several tests, let's split in more TravisCI builds depending on compilation options.

The new test matrix is 
```
    - WARPX_CI_REGULAR_CARTESIAN=TRUE
    - WARPX_CI_PSATD=TRUE
    - WARPX_CI_PYTHON_MAIN=TRUE
    - WARPX_CI_SINGLE_PRECISION=TRUE
    - WARPX_CI_RZ_OR_NOMPI=TRUE
    - WARPX_CI_QED=TRUE
```
where each element is made to have exactly 2 compilations, and should take ~30 min. Note the `OR` in `WARPX_CI_RZ_OR_NOMPI`: this will run all tests for which `RZ` is true or `noMPI` is true.

### CODE REVIEW

**Comment by RemiLehe** [comment]
Looks great! 
Since this relies on environment variables, would it make sense to prefix all environment variables by `WARPX_TEST_` (e.g. `WARPX_TEST_HAS_QED` instead of `HAS_QED`), so as to avoid any potential conflict?

**Comment by MaxThevenet** [comment]
Yeah makes sense, though I will try to keep their names shorter so that the build matrix is readable. Would prefix `TST_` work?

**Comment by MaxThevenet** [comment]
To avoid duplication, I defined a function to select tests. Please feel free to check that the logic is conserved :innocent:

**Comment by MaxThevenet** [comment]
@EZoni found a bug, I'll fix it and let you know.

**Comment by EZoni** [comment]
I think we should update the section "Testing the code" (subsection "Run the test suite locally") of the developers documentation, as the behavior of the script `run_test.sh` is slightly different now. When running tests that have non-empty compile options in the `addToCompileString` field (in the `WarpX-tests.ini` file), the corresponding CI option should be now set explicitly. For example, the command

`./run_test.sh pml_x_psatd`

which worked fine previously, must be now replaced by

`WARPX_CI_PSATD=TRUE ./run_test.sh pml_x_psatd`

otherwise the test cannot be found and turns out to be not valid.

I think it could be useful to also add a short table of correspondence between compile options and CI options: for example, the compile option `USE_PSATD=TRUE` corresponds to the CI option `WARPX_CI_PSATD=TRUE`, the compile option `USE_RZ=TRUE` corresponds to the CI option `WARPX_CI_DIM=RZ`, and so on.

Small related question about the options `PRECISION=FLOAT` and `USE_SINGLE_PRECISION_PARTICLES=TRUE`: are they both equivalent to `WARPX_CI_SINGLE_PRECISION=TRUE` now?

**Comment by MaxThevenet** [comment]
I think it's working!! Note that the longest run takes 38 min, but it will be shorten down when we won't need a separate compilation for electrostatic (this should be soon, and it will result in 2 compilations instead of 3. All other test builds have 2 compilations).

**Comment by MaxThevenet** [comment]
@EZoni Thanks for your comment. I think that, we the logic that we now used, we should recover the old behavior, e.g., `./run_test.sh pml_x_psatd` should work.

**Comment by MaxThevenet** [comment]
:innocent:

**Comment by MaxThevenet** [comment]
I rename `WARPX_CI_RZ_AND_NOMPI` to `WARPX_CI_RZ_OR_NOMPI` and added a GitHub workflow test to make sure the matrix runs each test **exactly once**. Should be ready now :).

**Comment by MaxThevenet** [comment]
Tests pass in ~30 min \o/. Anyone for merging this PR?

File: .github/workflows/source.yml [hunk 1]
```diff
@@ -25,6 +25,8 @@ jobs:
       run: .github/workflows/source/wrongFileNameInExamples
     - name: Examples are tested
       run: .github/workflows/source/inputsNotTested
+    - name: TravisCI matrix runs all tests exactly once
```
Comment (ax3l) [inline .github/workflows/source.yml]:
```suggestion
    - name: Check that the test matrix for Travis includes all tests
```

File: .github/workflows/source/test_travis_matrix.sh [hunk 2]
```diff
@@ -0,0 +1,27 @@
+#! /usr/bin/env sh
+
```
Comment (ax3l) [inline .github/workflows/source/test_travis_matrix.sh]:
```suggestion
set -eu -o pipefail
```

File: .github/workflows/source.yml (hunk 1 above)
Comment (ax3l) [inline .github/workflows/source.yml]:
```suggestion
    - name: Check that the test matrix for Travis includes all tests
```

File: .github/workflows/source/test_travis_matrix.sh [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+cp .github/workflows/source/travis_matrix.py Regression/
+cd Regression/
+
+# Put the name of all travis tests into a text file
+python prepare_file_travis.py && grep "\[" travis-tests.ini > travis_all_tests.txt
```
Comment (ax3l) [inline .github/workflows/source/test_travis_matrix.sh]: just make this two lines. the `set -eu -o ...` above will make sure the script fails if any of the commands fails

File: .github/workflows/source/travis_matrix.py [hunk 4]
```diff
@@ -0,0 +1,22 @@
+#! /usr/bin/env python
+
```
Comment (ax3l) [inline .github/workflows/source/travis_matrix.py]:
```suggestion
set -eu -o pipefail
```

Comment (MaxThevenet) [reply .github/workflows/source/travis_matrix.py]: that's a Python file.

Comment (ax3l) [reply .github/workflows/source/travis_matrix.py]: ha

**Pull request merged**